packages/flutter_tools/lib/src/android/gradle.dart

```diff
@@ -28,11 +29,44 @@ import '../reporting/reporting.dart';
 import 'android_sdk.dart';
 import 'android_studio.dart';
 
-final RegExp _assembleTaskPattern = RegExp(r'assemble(\S+)');
+/// Gradle utils in the current [AppContext].
+GradleUtils get gradleUtils => context.get<GradleUtils>() ?? GradleUtils.singleton();
+
+/// The implementation of Gradle Utils.
+@visibleForTesting
+class GradleUtils {
+  GradleUtils();
+
+  factory GradleUtils.singleton() {
```

Comment:
While we claim many things are singletons in our API docs, I don't think we've ever named a constructor `singleton` - it feels a bit leaky. Instead I would inject the GradleUtils in runInContext with the rest of the interfaces

Reply:
I think we also generally make a static public `instance` property for singletons, e.g. the various `Binding` classes

Reply:
that makes sense. I almost forgot about `runInContext`. Done